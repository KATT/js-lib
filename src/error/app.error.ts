/**
 * Base class for all our (not system) errors
 */
export class AppError extends Error {
  constructor (message?: string, public data?: any) {
    super(message)

    Object.defineProperty(this, 'name', {
      value: this.constructor.name,
    })

    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, this.constructor)
    } else {
      Object.defineProperty(this, 'stack', {
        value: new Error().stack,
      })
    }
  }
}
