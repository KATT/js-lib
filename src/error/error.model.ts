import { StringMap } from '../types'

/**
 * Extendable payload object to transfer custom additional data with AppError.
 */
export interface ErrorData extends StringMap<any> {
  /**
   * If present, will be displayed to the User as is.
   */
  userMessage?: string
}
