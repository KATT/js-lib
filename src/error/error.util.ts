import {
  AppError,
  ErrorData,
  ErrorObject,
  HttpErrorData,
  HttpErrorResponse,
  StringifyAnyOptions,
  _jsonParseIfPossible,
  _stringifyAny,
} from '..'

/**
 * Converts "anything" to ErrorObject.
 * Detects if it's HttpErrorResponse, HttpErrorObject, ErrorObject, Error, etc..
 * If object is Error - Error.message will be used.
 * Objects (not Errors) get converted to prettified JSON string (via `_stringifyAny`).
 */
export function _anyToErrorObject(o: any, opt?: StringifyAnyOptions): ErrorObject {
  if (o instanceof Error) {
    return _errorToErrorObject(o, opt?.includeErrorStack)
  }

  o = _jsonParseIfPossible(o)

  if (_isHttpErrorResponse(o)) {
    return o.error
  }

  if (_isErrorObject(o)) return o

  // Here we are sure it has no `data` property,
  // so, fair to return `data: {}` in the end
  // Also we're sure it includes no "error name", e.g no `Error: ...`,
  // so, fair to include `name: 'Error'`

  const message = _stringifyAny(o, opt)

  return {
    name: 'Error',
    message,
    data: {}, // empty
  }
}

export function _errorToErrorObject<DATA_TYPE extends ErrorData = ErrorData>(
  e: AppError<DATA_TYPE> | Error,
  includeErrorStack = false,
): ErrorObject<DATA_TYPE> {
  const obj: ErrorObject<DATA_TYPE> = {
    name: e.name,
    message: e.message,
    data: { ...(e as any).data }, // empty by default
  }

  if (includeErrorStack) {
    obj.stack = e.stack
  }

  return obj
}

export function _errorObjectToAppError<DATA_TYPE>(o: ErrorObject<DATA_TYPE>): AppError<DATA_TYPE> {
  const err = Object.assign(new AppError(o.message, o.data), {
    // name: err.name, // cannot be assigned to a readonly property like this
    // stack: o.stack, // also readonly e.g in Firefox
  })

  Object.defineProperty(err, 'name', {
    value: o.name,
    configurable: true,
  })

  Object.defineProperty(err, 'stack', {
    value: o.stack,
  })

  return err
}

export function _isHttpErrorResponse(o: any): o is HttpErrorResponse {
  return _isHttpErrorObject(o?.error)
}

export function _isHttpErrorObject(o: any): o is ErrorObject<HttpErrorData> {
  return (
    typeof o?.name === 'string' &&
    typeof o?.message === 'string' &&
    typeof o?.data?.httpStatusCode === 'number'
  )
}

export function _isErrorObject(o: any): o is ErrorObject {
  return (
    typeof o?.name === 'string' && typeof o?.message === 'string' && typeof o?.data === 'object'
  )
}
