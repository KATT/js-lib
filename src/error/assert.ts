import { ErrorData, HttpErrorData, _deepEquals, _stringifyAny } from '..'
import { AppError } from './app.error'

/**
 * Evaluates the `condition` (casts it to Boolean).
 * Expects it to be truthy, otherwise throws AppError.
 *
 * Should be used NOT for "expected" / user-facing errors, but
 * vice-versa - for completely unexpected and 100% buggy "should never happen" cases.
 *
 * It'll result in http 500 on the server (cause that's the right code for "unexpected" errors).
 * Pass { httpStatusCode: x } at errorData argument to override the http code (will be picked up by backend-lib).
 *
 * API is similar to Node's assert(), except:
 * 1. Throws js-lib's AppError
 * 2. Has a default message, if not provided
 * 3. Sets `userFriendly` flag to true, cause it's always better to have at least SOME clue, rather than fully generic "Oops" error.
 */
export function _assert(
  condition: any, // will be evaluated as Boolean
  message?: string,
  errorData?: Partial<HttpErrorData>,
): asserts condition {
  if (!condition) {
    throw new AssertionError(message || 'see stacktrace', {
      userFriendly: true,
      ...errorData,
    })
  }
}

/**
 * Like _assert(), but prints more helpful error message.
 * API is similar to Node's assert.equals().
 *
 * Does SHALLOW, but strict equality (===), use _assertDeepEquals() for deep equality.
 */
export function _assertEquals<T>(
  actual: any,
  expected: T,
  message?: string,
  errorData?: Partial<HttpErrorData>,
): asserts actual is T {
  if (actual !== expected) {
    const msg = [
      message || 'not equal',
      `got     : ${_stringifyAny(actual)}`,
      `expected: ${_stringifyAny(expected)}`,
    ]
      .filter(Boolean)
      .join('\n')

    throw new AssertionError(msg, {
      userFriendly: true,
      ...errorData,
    })
  }
}

/**
 * Like _assert(), but prints more helpful error message.
 * API is similar to Node's assert.deepEquals().
 *
 * Does DEEP equality via _deepEquals()
 */
export function _assertDeepEquals<T>(
  actual: any,
  expected: T,
  message?: string,
  errorData?: Partial<HttpErrorData>,
): asserts actual is T {
  if (!_deepEquals(actual, expected)) {
    const msg = [
      message || `not deeply equal`,
      `got     : ${_stringifyAny(actual)}`,
      `expected: ${_stringifyAny(expected)}`,
    ]
      .filter(Boolean)
      .join('\n')

    throw new AssertionError(msg, {
      userFriendly: true,
      ...errorData,
    })
  }
}

export class AssertionError extends AppError {
  constructor(message: string, data: ErrorData) {
    super(message, data)

    this.constructor = AssertionError
    ;(this as any).__proto__ = AssertionError.prototype
    Object.defineProperty(this, 'name', {
      value: this.constructor.name,
      configurable: true, // otherwise throws with "TypeError: Cannot redefine property: name"
    })

    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, this.constructor)
    } else {
      Object.defineProperty(this, 'stack', {
        value: new Error().stack, // eslint-disable-line unicorn/error-message
        configurable: true,
      })
    }
  }
}
