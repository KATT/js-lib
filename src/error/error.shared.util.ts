import { AppError, HttpError, HttpErrorData, objectUtil } from '..'
import { ErrorObject } from '../error/error.model'

const EMPTY_STRING_MSG = 'empty_string'

class ErrorSharedUtil {
  /**
   * Converts "anything" to String error message.
   * Objects get converted to prettified JSON string.
   */
  anyToErrorMessage (o: any): string {
    if (objectUtil.isObject(o)) {
      return JSON.stringify(o, null, 2)
    }

    return String(o) || EMPTY_STRING_MSG
  }

  anyToErrorObject (o: any): ErrorObject {
    if (o instanceof Error) return this.errorToErrorObject(o)

    return {
      name: 'Error',
      message: this.anyToErrorMessage(o),
      data: { ...(o || {}).data }, // empty by default
    }
  }

  errorToErrorObject (e: Error): ErrorObject {
    if (e instanceof AppError) return this.appErrorToErrorObject(e)

    return {
      name: e.name,
      message: e.message,
      stack: e.stack,
      data: { ...(e as any).data }, // empty by default
    }
  }

  errorObjectToAppError<T> (o: ErrorObject<T>): AppError<T> {
    const err = Object.assign(new AppError(o.message, o.data), {
      // name: err.name, // cannot be assigned to a readonly property like this
      stack: o.stack,
    })

    Object.defineProperty(err, 'name', {
      value: o.name,
    })

    return err
  }

  errorObjectToHttpError (o: ErrorObject): HttpError {
    const err = Object.assign(new HttpError(o.message, o.data), {
      // name: err.name, // cannot be assigned to a readonly property like this
      stack: o.stack,
    })

    Object.defineProperty(err, 'name', {
      value: o.name,
    })

    return err
  }

  appErrorToErrorObject<T> (err: AppError<T>): ErrorObject<T> {
    return {
      name: err.name,
      message: err.message,
      stack: err.stack,
      data: err.data,
    }
  }

  appErrorToHttpError (err: AppError<HttpErrorData>): HttpError {
    return new HttpError(err.message, err.data)
  }
}

export const errorSharedUtil = new ErrorSharedUtil()
