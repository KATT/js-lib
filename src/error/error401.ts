import { AppError } from './app.error'
import { ErrorData } from './error.model'

/**
 * HTTP 401: Unauthorized
 */
export class Error401 extends AppError {
  constructor (message?: string, data?: ErrorData) {
    super(message, data)

    Object.defineProperty(this, 'name', {
      value: this.constructor.name,
    })

    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, this.constructor)
    } else {
      Object.defineProperty(this, 'stack', {
        value: new Error().stack,
      })
    }
  }
}
