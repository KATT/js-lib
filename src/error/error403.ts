import { AppError } from './app.error'
import { ErrorData } from './error.model'

/**
 * HTTP 403: Forbidden
 */
export class Error403 extends AppError {
  constructor (message?: string, data?: ErrorData) {
    super(message, data)

    Object.defineProperty(this, 'name', {
      value: this.constructor.name,
    })

    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, this.constructor)
    } else {
      Object.defineProperty(this, 'stack', {
        value: new Error().stack,
      })
    }
  }
}
