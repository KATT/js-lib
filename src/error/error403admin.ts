import { ErrorData } from './error.model'
import { Error403 } from './error403'

/**
 * HTTP 403: Admin access forbidden
 */
export class Error403Admin extends Error403 {
  constructor (message?: string, data?: ErrorData) {
    super(message, data)

    Object.defineProperty(this, 'name', {
      value: this.constructor.name,
    })

    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, this.constructor)
    } else {
      Object.defineProperty(this, 'stack', {
        value: new Error().stack,
      })
    }
  }
}
