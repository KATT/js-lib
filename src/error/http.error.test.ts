import { _errorToErrorObject } from '../index'
import { HttpError } from './http.error'

const throwHttpError = () => {
  throw new HttpError('error', { httpStatusCode: 500 })
}
const throwHttpErrorAsync = async () => {
  throw new HttpError('error', { httpStatusCode: 500 })
}

test('default error to match snapshot', async () => {
  const err = new HttpError('error', { httpStatusCode: 500 })
  // console.log(err)
  expect(_errorToErrorObject(err, true)).toMatchSnapshot({
    stack: expect.stringContaining('HttpError'),
  })

  expect(throwHttpError).toThrow(HttpError)
  await expect(throwHttpErrorAsync()).rejects.toThrow(HttpError)
})
