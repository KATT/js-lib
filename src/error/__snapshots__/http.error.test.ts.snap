// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default error to match snapshot 1`] = `
Object {
  "data": Object {
    "httpStatusCode": 500,
  },
  "message": "error",
  "name": "HttpError",
  "stack": StringContaining "HttpError",
}
`;
